consensus/consensus_v2.go

```diff
@@ -290,168 +298,168 @@ func (consensus *Consensus) BlockCommitSigs(blockNum uint64) ([]byte, error) {
 
 // Start waits for the next new block and run consensus
 func (consensus *Consensus) Start(
-	blockChannel chan *types.Block, stopChan, stoppedChan, startChannel chan struct{},
+	stopChan chan struct{},
 ) {
 	go func() {
-		toStart := make(chan struct{}, 1)
-		isInitialLeader := consensus.IsLeader()
-		if isInitialLeader {
-			consensus.getLogger().Info().Time("time", time.Now()).Msg("[ConsensusMainLoop] Waiting for consensus start")
-			// send a signal to indicate it's ready to run consensus
-			// this signal is consumed by node object to create a new block and in turn trigger a new consensus on it
-			go func() {
-				<-startChannel
-				toStart <- struct{}{}
-				consensus.getLogger().Info().Time("time", time.Now()).Msg("[ConsensusMainLoop] Send ReadySignal")
-				consensus.ReadySignal <- SyncProposal
-			}()
-		}
 		consensus.getLogger().Info().Time("time", time.Now()).Msg("[ConsensusMainLoop] Consensus started")
-		defer close(stoppedChan)
-		ticker := time.NewTicker(250 * time.Millisecond)
-		defer ticker.Stop()
+		go func() {
+			ticker := time.NewTicker(250 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-stopChan:
+					return
+				case <-ticker.C:
+					consensus.mutex.Lock()
+					consensus.tick()
```

Comment:
I think if you move mutex inside tick() function would be better. Because there you can use defer to unlock mutex. So, for any reason if tick() function fails, mutex would be unlocked. In current implementation if tick() fails, mutex will remain locked.

Reply:
fixed